vowpalwabbit/cli_help_formatter.cc

```diff
@@ -0,0 +1,187 @@
+
+    void visit(typed_option<uint32_t>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<uint64_t>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<int64_t>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<int32_t>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<bool>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<float>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<std::string>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", option.default_value()); }
+    };
+    void visit(typed_option<std::vector<std::string>>& option) override
+    {
+      if (option.default_value_supplied()) { m_ret_val = fmt::format("{}", fmt::join(option.default_value(), ", ")); }
+    };
+  };
+
+  extractor ex(ret_val);
+  opt.accept(ex);
+  return ret_val;
+}
+
+std::vector<std::string> split_string_by_newline(const std::string& str)
+{
+  auto result = std::vector<std::string>{};
+  auto ss = std::stringstream{str};
+
+  for (std::string line; std::getline(ss, line, '\n');) result.push_back(line);
+
+  return result;
+}
+
+constexpr size_t LEFT_COL_WIDTH = 40;
+constexpr size_t RIGHT_COL_WIDTH = 60;
+static const std::string INDENT = "    ";
+
+std::string cli_help_formatter::format_help(const std::vector<option_group_definition>& groups)
+{
+  std::stringstream overall_ss;
+  for (const auto& group : groups)
+  {
+    overall_ss << group.m_name << ":\n";
+    for (const auto& option : group.m_options)
+    {
+      if (option->m_hidden_from_help) { continue; }
+      std::stringstream ss_option_name;
+      if (!option->m_short_name.empty()) { ss_option_name << "-" << option->m_short_name << ", "; }
+      ss_option_name << "--" << option->m_name;
+      auto type = type_string(*option);
+      if (type == "list[str]") { ss_option_name << " args..."; }
+      else if (type != "bool")
+      {
+        ss_option_name << " arg";
+      }
+      std::string option_name_str = ss_option_name.str();
+
+      std::stringstream ss_description;
+      ss_description << option->m_help;
+
+      ss_description << " (type: " << type_string(*option);
+
+      if (!default_value_string(*option).empty()) { ss_description << ", default: " << default_value_string(*option); }
+      if (option->m_keep) { ss_description << ", keep"; }
```

Comment:
May want to put one_of after default. Wondering if users will know what `keep` and `necessary` mean.

Reply:
One of is after default. I think keep and necessary are important pieces of info we need to a better job at explaining. They allow you to understand how the command line is interpreted and stored

Reply:
I was suggesting putting one_of after default and before `keep` and `necessary` since `type`, `default` and `one_of` are all related to what a user can input. And yea I agree keep and necessary are important, just wasn't sure it would be clear what they meant. If you release a doc later when you're done refactoring something like `keep: persist this value when saving and loading a model` would be useful

Reply:
Oh I see what you're saying yeah that makes sense